### PR TITLE
mtd/*: drop .write() if .write_page() is implemented

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

The old `.write()` function is only used as a fall-back if `.write_page()` is not implemented.
We can drop it.

### Testing procedure

Everything will fall back to `.write_page` now, so no user visible change.
On the driver level, `.write_page` can be much simpler than `.write` as all the common logic is handled by MTD and we can do partial writes (only write to page boundary, MTD will split up the transaction).

 - [x] `native` (tested by CI)
 - [x] ESP8266/ESP32
 - [x] AT25xxx
 - [x] MTD Mapper (tested by CI)
 - [x] MTD SDcard

<details><summary>AT25xxx</summary>

#### tests/mtd_raw
```
2023-03-09 23:26:41,366 # init MTD_0… OK (128 kiB)
2023-03-09 23:26:43,208 # > test 0
2023-03-09 23:26:43,208 # [START]
2023-03-09 23:26:43,286 # [SUCCESS]

2023-03-09 23:28:20,306 # > read 0 0 32
2023-03-09 23:28:20,312 # 00000000  31  32  33  34  35  36  FF  F7  6C  69  74  74  6C  65  66  73
2023-03-09 23:28:20,318 # 00000010  2F  E0  00  10  00  00  02  00  00  01  00  00  00  02  00  00

2023-03-09 23:28:52,800 # > read 0 0 32
2023-03-09 23:28:52,806 # 00000000  31  61  62  63  64  65  66  67  68  69  74  74  6C  65  66  73
2023-03-09 23:28:52,812 # 00000010  2F  E0  00  10  00  00  02  00  00  01  00  00  00  02  00  00
```
</details>

<details><summary>ESP32</summary>

#### tests/mtd_raw
```
2023-03-09 23:31:24,456 # > test 0
2023-03-09 23:31:24,457 # [START]
2023-03-09 23:31:24,603 # [SUCCESS]

2023-03-09 23:32:43,227 # > read 0 0 32
2023-03-09 23:32:43,234 # 00000000  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
2023-03-09 23:32:43,240 # 00000010  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF

2023-03-09 23:32:48,918 # > write 0 1 abcdefgh
2023-03-09 23:32:50,963 # 00000000  FF  61  62  63  64  65  66  67  68  FF  FF  FF  FF  FF  FF  FF
2023-03-09 23:32:50,969 # 00000010  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF

2023-03-09 23:32:50,957 # > read 0 0 32
2023-03-09 23:32:50,963 # 00000000  FF  61  62  63  64  65  66  67  68  FF  FF  FF  FF  FF  FF  FF
2023-03-09 23:32:50,969 # 00000010  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF  FF
```
</details>

<details><summary>MTD SDcard</summary>

#### tests/mtd_raw
```
2023-12-13 16:55:18,923 - INFO # main(): This is RIOT! (Version: 2024.01-devel-369-ga45d2-mtd_drop_write)
2023-12-13 16:55:18,923 - INFO # Manual MTD test
2023-12-13 16:55:19,034 - INFO # init MTD_0… OK (15223 MiB)
test 0
2023-12-13 16:55:26,746 - INFO # > test 0
2023-12-13 16:55:26,747 - INFO # 
2023-12-13 16:55:26,747 - INFO # [START]
2023-12-13 16:55:27,706 - INFO # [SUCCESS]
```
</details>

<details><summary>MTD spi_nor, at24mac, MTD SDMMC (with #20180)</summary>

#### tests/mtd_raw
```
2023-12-13 17:07:07,125 - INFO # Manual MTD test
2023-12-13 17:07:07,128 - INFO # init MTD_0… OK (8 MiB)
2023-12-13 17:07:07,131 - INFO # init MTD_1… OK (256 byte)
2023-12-13 17:07:07,342 - INFO # init MTD_2… OK (15223 MiB)
test 0
2023-12-13 17:07:10,339 - INFO # > test 0
2023-12-13 17:07:10,339 - INFO # 
2023-12-13 17:07:10,339 - INFO # [START]
2023-12-13 17:07:11,608 - INFO # [SUCCESS]
test 1
2023-12-13 17:07:14,593 - INFO # > test 1
2023-12-13 17:07:14,593 - INFO # 
2023-12-13 17:07:14,594 - INFO # [START]
2023-12-13 17:07:14,648 - INFO # [SUCCESS]
test 2
2023-12-13 17:07:16,275 - INFO # > test 2
2023-12-13 17:07:16,275 - INFO # 
2023-12-13 17:07:16,275 - INFO # [START]
2023-12-13 17:07:16,963 - INFO # [SUCCESS]
```
</details>

### Issues/PRs references

includes #15378
probably also needs #14953 and #19258

### CODE REVIEW

**Comment by jue89** [comment]
Do you an idea how test this without owning the respective hardware?

Or can we prove that `mtd`'s translation works in 100% of all cases?

**Comment by benpicco** [comment]
I don't think we have to test this on *every* implementation, the functionality is the same everywhere.
However, we should make sure nothing breaks when layering mtd implementations, e.g. with mtd_mapper.

I didn't do any tests yet, this is more a PoC, but the idea is to only have one write implementation per MTD backend.

**Comment by bergzand** [comment]
Please rebase!

**Comment by benpicco** [comment]
rebased

**Comment by MrKevinWeiss** [comment]
Any blockers for this, does it just need some testing?

**Comment by fjmolinas** [comment]
> I don't think we have to test this on _every_ implementation, the functionality is the same everywhere.
> However, we should make sure nothing breaks when layering mtd implementations, e.g. with mtd_mapper.
> 
> I didn't do any tests yet, this is more a PoC, but the idea is to only have one write implementation per MTD backend.

Las tcomment sated this was a PoC, is this still the case?

**Comment by benpicco** [comment]
If `.write_page()` is implemented `.write` will [not be used](https://github.com/RIOT-OS/RIOT/blob/master/drivers/mtd/mtd.c#L185), so we can drop it and save some ROM.

**Comment by fjmolinas** [comment]
Do you have a test procedure for this?

**Comment by fjmolinas** [comment]
> If `.write_page()` is implemented `.write` will [not be used](https://github.com/RIOT-OS/RIOT/blob/master/drivers/mtd/mtd.c#L185), so we can drop it and save some ROM.

I only worry that eventually, some external code could be relying on this function, so at least it needs to be tagged as removal. We could also conditionally include (with a pseudo module) and then remove it in a release or two, but maybe it's not worth the effort. Let's just make sure it will be tagged in release notes.

**Comment by benpicco** [comment]
`tests/mtd_raw`contains an extensive `test` command that can be used to verify MTD functionality

File: tests/unittests/tests-mtd/tests-mtd.c [hunk 1]
```diff
@@ -206,19 +206,19 @@ static void test_mtd_write_read(void)
     /* out of bounds write (addr) */
     ret = mtd_write(dev, buf, dev->pages_per_sector * dev->page_size * dev->sector_count,
                     sizeof(buf));
-    TEST_ASSERT_EQUAL_INT(-EOVERFLOW, ret);
```
Comment (fjmolinas) [inline tests/unittests/tests-mtd/tests-mtd.c:209]: `mtd_flashpage` still uses `write` so this unittest would not work in that case would it?

**Comment by OlegHahm** [comment]
@fjmolinas, @benpicco, where are we at here?

File: drivers/mtd_sdcard/mtd_sdcard.c [hunk 2]
```diff
@@ -85,22 +85,6 @@ static int mtd_sdcard_read_page(mtd_dev_t *dev, void *buff, uint32_t page,
...
-    sd_rw_response_t err;
-    sdcard_spi_write_blocks(mtd_sd->sd_card, addr / SD_HC_BLOCK_SIZE,
-                            buff, SD_HC_BLOCK_SIZE,
-                            size / SD_HC_BLOCK_SIZE, &err);
-
-    if (err == SD_RW_OK) {
-        return 0;
-    }
-    return -EIO;
-}
```
Comment (fjmolinas) [inline drivers/mtd_sdcard/mtd_sdcard.c:102]: This function def needs to be removed since the writes of less than SD_HC_BLOCK_SIZE are not possible.

**Review by fjmolinas — COMMENTED** [review]
Re-visiting this I have some questions

> The old `.write()` function is only used as a fall-back if `.write_page()` is not implemented.
We can drop it.

`mtd_write_page_raw` is supposed to not perform automatic `read-modify-write` cycles,  but on the other end in many cases `.write_page` does perform `read-modify-write` cycles depending on the implementation, or drops.

- mtd_sdcard: will return an error if writing not aligned to a block
- mtd_mci: uses a non memory safe buffer to perform a read-write-modify-cycle
- mtd_flashpage (uses write): will error if non block aligned.

**Comment by benpicco** [comment]
#19258 fixes the issue with `mtd_flashpage`.

As for `mtd_mci` we rely on the upper layer to prevent concurrent access, this is usually done by the filesystem but if needed it could be moved to MTD.

`mtd_sdcard` would also return an error before this on unaligned access, so the only change is that unaligned access now *always* works if the `mtd_write_pagge` module is used.

**Comment by kfessel** [comment]
please rebase so this pr contains every `mtd_dev_t` has the `.write_page()` (or just master)

**Comment by kfessel** [comment]
Some time ago you did mention you did not test this, did you do now?  -- in case which backends?

**Comment by benpicco** [comment]
bors try

**Comment by benpicco** [comment]
All backends are tested now, the virtual backends are tested by CI.

**Review by maribu — APPROVED** [review]

**Comment by benpicco** [comment]
Thank you! :smiley:

**Pull request merged**